### PR TITLE
Make seiscomp3 compatible with Mac OS X

### PR DESCRIPTION
**Pull request by gilcel**
This commit allows compiling of seiscomp3-jakarta (2016) on Mac OS X.

All the seiscomp3-plugins are now enabled and compiles correctly under OS X.
Even the seedlink locator-plugin, which did not work due to lack of open_memstream on OS X, is now supported.
# Seiscomp3 for Mac OS X installation notes
## Pre-Requirements (Xcode, GFortran, homebrew)

Compiled on Mac OS X El Capitan 10.11.4 with Xcode 7.3 and gfortran 5.2
### Install Xcode from command-line

Open up Terminal.app and type:
 `xcode-select --install`
### Install GFortran

https://gcc.gnu.org/wiki/GFortranBinaries
### Install Homebrew

`/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"`
Webpage here:
    http://brew.sh

All the packages will be installed in /usr/local/Cellar

Check if your system is correctly setup with:
    `brew doctor`
## Install seiscomp3 dependencies

`brew install cmake boost qt4 md5sha1sum`
### Force link openssl & flex

We need to force link openssl & flex so that they will be available in /usr/local/lib/ 
`brew install openssl`
`brew link --force openssl`
`brew install flex`
`brew link --force flex`
### Install MySQL

`brew install mysql`

Copy default MYSQL configuration file to /usr/local/etc/
`sudo cp $(brew --prefix mysql)/support-files/my-default.cnf /etc/my.cnf`
## Compile seiscomp3 for OS X

```
cd ~/Downloads/seiscomp3-git/
mkdir build-seiscomp3-osx
cd build-seiscomp3-osx
cmake -DCMAKE_INSTALL_PREFIX=${HOME}/seiscomp3 ../seiscomp3-src-osx
make
make install
```

If everything compiled fine, the files will be installed in ${HOME}/seiscomp3
## Increase max open files for seedlink

To avoid getting seedlink errors when starting seiscomp with "files open exceed max files ..." we need to increase the max open files, we need to set it at startup.

Place this `limit.seiscomp3-maxfiles.plist` file in `/Library/LaunchDaemons/`

`<?xml version="1.0" encoding="UTF-8"?>  
    <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"  
       "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
    <plist version="1.0">  
    <dict>
    <key>Label</key>
    <string>limit.seiscomp3-maxfiles</string>
    <key>ProgramArguments</key>
    <array>
     <string>launchctl</string>
     <string>limit</string>
     <string>maxfiles</string>
     <string>524288</string>
     <string>524288</string>
   </array>
   <key>RunAtLoad</key>
   <true/>
   <key>ServiceIPC</key>
   <false/>
 </dict>
</plist>  
`

Set permission:
`sudo chown root:wheel /Library/LaunchDaemons/limit.seiscomp3-maxfiles.plist`
Launch it:
`sudo launchctl load -w /Library/LaunchDaemons/limit.seiscomp3-maxfiles.plist`

### CODE REVIEW

**Comment by gempa-jabe** [comment]
Thanks for the substantial patch. It will take a closer look to check all your changes. I will comment in the commit itself.

File: CMakeLists.txt [hunk 1]
```diff
@@ -39,11 +39,7 @@ EXEC_PROGRAM(
 )
 
 IF (SC_GLOBAL_PYTHON_WRAPPER)
-	IF (APPLE)
```
Comment (gempa-jabe) [inline CMakeLists.txt]: Why did you remove it? Didn't it work? We had to introduce that for our OSX builds to avoid conflicts if you have a homebrew Python version installed. See Modules/FindPythonLIbsApple.cmake.

File: CMakeLists.txt [hunk 2]
```diff
@@ -101,6 +97,39 @@ IF (CMAKE_COMPILER_IS_GNUCC)
 	ENDIF ( ${CMAKE_BUILD_TYPE} MATCHES "Debug" )
 ENDIF (CMAKE_COMPILER_IS_GNUCC)
 
+
+# Set some specific settings for Mac OS X - Darwin
+#
+IF (MACOSX)
+    
+    # For MAC OS X we set the GUI to ON for ssconfig etc.
+    SET(SC_GLOBAL_GUI ON)
```
Comment (gempa-jabe) [inline CMakeLists.txt]: That is a configuration flag that should be set by the user independently of the target platform. Even on a Mac you can build without GUI if you want. I don't see a reason to remove that flexibility.

File: src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp [hunk 3]
```diff
@@ -217,7 +217,7 @@ bool ConnectionDialog::fetchDatabase() {
 		Util::StopWatch timer;
 		while ( timer.elapsed() < TimeSpan(5.0) ) {
 			(*_connection)->poll();
-			while ( msg = (*_connection)->readMessage(false) ) {
+			while ( msg == (*_connection)->readMessage(false) ) {
```
Comment (gempa-jabe) [inline src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp]: That is definitely wrong and will break the code. What is meant here is:

```c++
while ( (msg = (*_connection)->readMessage(false)) != NULL )
```

or more readable

```c++
while ( true ) {
  msg = (*_connection)->readMessage(false);
  if ( msg == NULL ) break;
  ...
}
```

Did clang raise an error or just a warning?

File: src/sed/plugins/locator/nll/CMakeLists.txt [hunk 4]
```diff
@@ -11,14 +11,32 @@ FILE(GLOB LIBNLL_ALOMAX_OCTTREE_SOURCES ${NLL_SOURCE_DIR}/octtree/*.c)
...
 FILE(GLOB LIBNLL_ALOMAX_VECTOR_SOURCES ${NLL_SOURCE_DIR}/vector/*.c)
 
+IF(MACOSX)
+FILE(GLOB LIBNLL_OPEN_MEMSTREAM_OSX_SOURCES ${NLL_SOURCE_DIR}/open_memstream-osx/*.c)
+ENDIF(MACOSX)
+
 IF(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
 	IF (CMAKE_COMPILER_IS_GNUCC)
-		ADD_DEFINITIONS(-D_GNU_SOURCE)
+	ADD_DEFINITIONS(-D_GNU_SOURCE)
```
Comment (gempa-jabe) [inline src/sed/plugins/locator/nll/CMakeLists.txt]: You don't like the indentation? ;)

Comment (gilcel) [inline src/sed/plugins/locator/nll/CMakeLists.txt]: Oh sorry! I was trying something here ;-)

File: src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp (hunk 3 above)
Comment (gilcel) [inline src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp]: Indeed: clang gave a Warning. But the Warning is right, isn't it ?

File: src/sed/plugins/locator/nll/CMakeLists.txt [hunk 5]
```diff
@@ -27,20 +45,34 @@ SET( ${PACKAGE_NAME}_SOURCES
 	${LIBNLL_ALOMAX_OCTTREE_SOURCES}
 	${LIBNLL_ALOMAX_RAN1_SOURCES}
 	${LIBNLL_ALOMAX_VECTOR_SOURCES}
+    ${LIBNLL_OPEN_MEMSTREAM_OSX_SOURCES}
 	nll_locator.cpp
 )
+ENDIF(MACOSX)
```
Comment (gempa-jabe) [inline src/sed/plugins/locator/nll/CMakeLists.txt]: I would prefer to just add the MACOSX specific files, e.g.:

```cmake
IF(MACOSX)
  SET(${PACKAGE_NAME}_SOURCES
    ${PACKAGE_NAME}_SOURCES
    ${LIBNLL_OPEN_MEMSTREAM_OSX_SOURCES}
  )
ENDIF(MACOSX)
```

Then we don't need an additional Linux block.

File: src/sed/plugins/locator/nll/CMakeLists.txt [hunk 6]
```diff
@@ -27,20 +45,34 @@ SET( ${PACKAGE_NAME}_SOURCES
...
 
+IF(MACOSX)
+INCLUDE_DIRECTORIES(${NLL_SOURCE_DIR}/open_memstream-osx)
+ENDIF(MACOSX)
+
 SC_ADD_PLUGIN_LIBRARY(${PACKAGE_NAME} locnll "")
 SC_LINK_LIBRARIES_INTERNAL(locnll core)
 
+IF(MACOSX)
+SC_LINK_LIBRARIES_INTERNAL(locnll core)
```
Comment (gempa-jabe) [inline src/sed/plugins/locator/nll/CMakeLists.txt]: Doesn't that duplicate line 71?

File: CMakeLists.txt (hunk 2 above)
Comment (gilcel) [inline CMakeLists.txt]: Ok, I will remove it then. The user must then set the flag manually with cmake.

File: src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp (hunk 3 above)
Comment (gempa-jabe) [inline src/gui-qt4/libs/seiscomp3/gui/core/connectiondialog.cpp]: Yes, the warning is right. We have to enclose the condition with parenthesis. I will do that in master because it is a general issue.

File: CMakeLists.txt (hunk 1 above)
Comment (gilcel) [inline CMakeLists.txt]: Correction: I tried with the initial CMakeLists.txt on OS X: while cmake 3.5.2 finds the PythonLibs, I get the following error when invoking "make":
`    [  0%] Building CXX object src/system/libs/swig/CMakeFiles/_Config.dir/ConfigPYTHON_wrap.cxx.o
    /src/system/libs/swig/ConfigPYTHON_wrap.cxx:155:11: fatal error: 'Python.h'
        file not found
        #include <Python.h>
        1 error generated.
    make[2]: *** [src/system/libs/swig/CMakeFiles/_Config.dir/ConfigPYTHON_wrap.cxx.o] Error 1
    make[1]: *** [src/system/libs/swig/CMakeFiles/_Config.dir/all] Error 2
    make: *** [all] Error 2
`

So should I leave use my code again ?

Comment (gempa-jabe) [inline CMakeLists.txt]: OK, where is your Pyhton.h? Maybe we don't set the correct include directory. Can you check in CMakeCache.txt the value of PYTHON_INCLUDE_DIR? For debugging you can also change into "build/src/system/libs/swig" and run "VERBOSE=1 make" to see the compiler call. It should include the "-I" for the Python include directory.

Comment (gilcel) [inline CMakeLists.txt]: Here's is the content of CMakeCache.txt when using your code to find Python libs (note: Python of OS X, no Homebrew Python installed)
//Path to a program.
PYTHON_CONFIG_EXECUTABLE:FILEPATH=/usr/bin/python-config

//Path to a program.
PYTHON_EXECUTABLE:FILEPATH=/usr/bin/python

//Path to where Python.h is found
PYTHON_INCLUDE_DIR:PATH=/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7

//Path to where libpython2.7.dylib is found
PYTHON_LIBRARY:PATH=/System/Library/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib

Comment (gilcel) [inline CMakeLists.txt]: `/src/system/libs/swig]>VERBOSE=1 make`

Building CXX object src/system/libs/swig/CMakeFiles/_Config.dir/ConfigPYTHON_wrap.cxx.o
cd /Users/gilles/Development/seiscomp3-git/build-seiscomp3-osx-jakarta-2016/src/system/libs/swig && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -DSC_ALL_SHARED -D_Config_EXPORTS -D_REENTRANT -I/Users/gilles/Development/seiscomp3-git/seiscomp3-src-osx-jakarta-2016/src/system/libs -I/Users/gilles/Development/seiscomp3-git/build-seiscomp3-osx-jakarta-2016/src/system/libs -I/Users/gilles/Development/seiscomp3-git/seiscomp3-src-osx-jakarta-2016/src/system/libs/swig  -fPIC   -DSHARED_LIBRARY_SUFFIX="\".dylib\"" -DSHARED_MODULE_SUFFIX="\".so\"" -o CMakeFiles/_Config.dir/ConfigPYTHON_wrap.cxx.o -c /Users/gilles/Development/seiscomp3-git/seiscomp3-src-osx-jakarta-2016/src/system/libs/swig/ConfigPYTHON_wrap.cxx
/Users/gilles/Development/seiscomp3-git/seiscomp3-src-osx-jakarta-2016/src/system/libs/swig/ConfigPYTHON_wrap.cxx:155:11: fatal error: 'Python.h'
      file not found
#include <Python.h>
          ^
1 error generated.
make[2]: *** [src/system/libs/swig/CMakeFiles/_Config.dir/ConfigPYTHON_wrap.cxx.o] Error 1
make[1]: *** [src/system/libs/swig/CMakeFiles/_Config.dir/all] Error 2
make: *** [all] Error 2
[gilles@sangoku /Users/gilles/Development/seiscomp3-git/build-seiscomp3-osx-jakarta-2016/src/system/libs/swig]>

Comment (gempa-jabe) [inline CMakeLists.txt]: Can you confirm that ```/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/Python.h``` exists and that /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 is not in the default include search path? According to https://github.com/SeisComP3/seiscomp3/blob/master/src/system/libs/swig/CMakeLists.txt#L1 we do set the Python include directory.

Comment (gilcel) [inline CMakeLists.txt]: Yes it exists, here:
ls -l /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/Python.h
-rw-r--r--  1 root  wheel  4329 Aug 23  2015 /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/Python.h

`cat /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7/Python.h`

#ifndef Py_PYTHON_H
#define Py_PYTHON_H
